core/src/databases_store/gcs.rs

```diff
@@ -0,0 +1,214 @@
+    }
+
+    pub async fn write_rows_to_csv(
+        table: &Table,
+        schema: &TableSchema,
+        rows: &Vec<Row>,
+    ) -> Result<(), anyhow::Error> {
+        let field_names = schema
+            .columns()
+            .iter()
+            .map(|c| c.name.clone())
+            .collect::<Vec<_>>();
+
+        // Read all rows and upload to GCS
+        let mut wtr = Writer::from_writer(vec![]);
+
+        // Write the header row.
+        wtr.write_record(field_names.as_slice())?;
+        for row in rows.iter() {
+            wtr.write_record(row.to_csv_record(&field_names)?.as_slice())?;
+        }
+
+        let csv = wtr.into_inner()?;
+
+        Object::create(
+            &Self::get_bucket()?,
+            csv,
+            &Self::get_csv_storage_file_path(table),
+            "text/csv",
+        )
+        .await?;
+
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl DatabasesStore for GoogleCloudStorageDatabasesStore {
+    async fn load_table_row(&self, table: &Table, row_id: &str) -> Result<Option<Row>> {
+        let rows = Self::get_rows_from_csv(table).await?;
+        let row = rows.iter().find(|r| r.row_id == row_id);
+        Ok(row.cloned())
+    }
+
+    async fn list_table_rows(
+        &self,
+        table: &Table,
+        limit_offset: Option<(usize, usize)>,
+    ) -> Result<(Vec<Row>, usize)> {
+        let rows = Self::get_rows_from_csv(table).await?;
+        match limit_offset {
+            Some((limit, offset)) => {
+                let total = rows.len();
+                let rows = rows.into_iter().skip(offset).take(limit).collect();
+                Ok((rows, total))
+            }
+            None => {
+                let total = rows.len();
+                Ok((rows, total))
+            }
+        }
+    }
+
+    async fn batch_upsert_table_rows(
+        &self,
+        table: &Table,
+        rows: &Vec<Row>,
+        truncate: bool,
+    ) -> Result<()> {
+        let mut now = utils::now();
+        let mut new_rows = rows.clone();
+        let mut merge_rows_duration = 0;
+
+        // We need to merge the existing rows with the new rows based on the row_id.
+        if !truncate {
+            // We download all the rows from the CSV and merge them with the new rows.
+            // This is not super efficient if we get rows one by one but it's simple and works.
+            // Non-truncate upserts are < 4% of our total upserts.
+
+            let previous_rows = Self::get_rows_from_csv(table).await?;
+
+            // Use Hashmaps with the row_id as the key.
+            let previous_rows_map = previous_rows
+                .into_iter()
+                .map(|r| (r.row_id.clone(), r.clone()))
+                .collect::<HashMap<_, _>>();
+
+            let new_rows_map = new_rows
+                .into_iter()
+                .map(|r| (r.row_id.clone(), r.clone()))
+                .collect::<HashMap<_, _>>();
+
+            // Merge the two maps to get the final rows.
```

Comment:
nit: I'd point out in the comment that the new one wins over the old in the merge (iiuc)